src/som/interpreter/nodes/literals/ArrayLiteralNode.java

```diff
@@ -1,81 +1,158 @@
-    return converted;
+  private static final class Doubles extends Specialized {
+    private Doubles(final ExpressionNode[] expressions, final SourceSection source) {
+      super(expressions, source);
+    }
+
+    @Override
+    protected Object executeSpecialized(final VirtualFrame frame) {
+      return ArraySetAllStrategy.evalForRemaining(frame, expressions,
+          new double[expressions.length], SArray.FIRST_IDX);
+    }
+
+    @Override
+    protected boolean expectedType(final SMutableArray arr) {
+      return arr.isDoubleType();
+    }
   }
 
-  @ExplodeLoop
-  private boolean[] convertToBooleanArray(final Object[] evaluated) {
-    boolean[] converted = new boolean[evaluated.length];
-    for (int i = 0; i < evaluated.length; i++) { converted[i] = (boolean) evaluated[i]; }
-    return converted;
+  private static final class Longs extends Specialized {
+    private Longs(final ExpressionNode[] expressions, final SourceSection source) {
+      super(expressions, source);
+    }
+
+    @Override
+    protected Object executeSpecialized(final VirtualFrame frame) {
+      return ArraySetAllStrategy.evalForRemaining(frame, expressions,
+          new long[expressions.length], SArray.FIRST_IDX);
+    }
+
+    @Override
+    protected boolean expectedType(final SMutableArray arr) {
+      return arr.isLongType();
+    }
   }
 
-  private Object executeAndSpecialize(final VirtualFrame frame) {
-    Object[] evaluated = evaluateAsArray(frame);
+  private static final class Empty extends Specialized {
+    private Empty(final ExpressionNode[] expressions, final SourceSection source) {
+      super(expressions, source);
+    }
 
-    if (headSameTypeAsTail(evaluated)) {
-      if (evaluated[0] instanceof Long) {
-        return convertToLongArray(evaluated);
-      } else if (evaluated[0] instanceof Double) {
-        return convertToDoubleArray(evaluated);
-      } else if (evaluated[0] instanceof Boolean) {
-        return convertToBooleanArray(evaluated);
-      }
+    @Override
+    protected Object executeSpecialized(final VirtualFrame frame) {
+      return ArraySetAllStrategy.evalForRemainingNils(frame, expressions,
+          SArray.FIRST_IDX);
     }
 
-    return evaluated;
+    @Override
+    protected boolean expectedType(final SMutableArray arr) {
+      return arr.isEmptyType();
+    }
   }
 
-  @Override
-  public Object executeGeneric(final VirtualFrame frame) {
-    return new SMutableArray(executeAndSpecialize(frame), Classes.arrayClass);
+  private static final class Objects extends Specialized {
+    private Objects(final ExpressionNode[] expressions, final SourceSection source) {
+      super(expressions, source);
+    }
+
+    @Override
+    protected Object executeSpecialized(final VirtualFrame frame) {
+      return ArraySetAllStrategy.evalForRemaining(frame, expressions,
+          new Object[expressions.length], SArray.FIRST_IDX);
```

Comment:
That's an artifact of my work on metaobject protocols, for those, I put some hidden things into arrays, then it wasn't 0. At the moment it is, but I kept it around for the possibility that I might need it one day. Not sure it's a great idea.